frontend/src/scenes/debug/NotebookTest.tsx

```diff
@@ -0,0 +1,210 @@
+            } catch (e) {
+                lemonToast.error(`Unable to compile bytecode: ${e}`)
+                console.error(e)
+                return
+            }
+            breakpoint()
+            // update the bytecode
+            if (response && 'bytecode' in response && Array.isArray(response.bytecode)) {
+                actions.updateBytecode(response.bytecode, fetchedUntil)
+            }
+            // set as
+        },
+        runFrom: async ({ index }) => {
+            if (index > values.bytecodeUntil) {
+                return
+            }
+            const bytecode = values.bytecode
+            let stdout: string[] = []
+
+            let nextState: any[] | VMState = bytecode
+
+            if (index > 0 && values.execResults[index - 1]) {
+                // add the new bytecode even if the rest of the state is old, the compiled section should not have changed
+                nextState = { ...values.execResults[index - 1].state, bytecode: bytecode }
+            }
+
+            while (true) {
+                const response = execHog(nextState, {
+                    functions: {
+                        print: (...args) => {
+                            stdout.push(args.map((a) => printHogValue(a)).join(' '))
+                        },
+                    },
+                    asyncFunctions: {
+                        __nbbb__: async () => true, // not actually a function
+                    },
+                })
+                if (response.finished) {
+                    break
+                }
+                if (
+                    response.asyncFunctionName === '__nbbb__' &&
+                    Array.isArray(response.asyncFunctionArgs) &&
+                    response.state
+                ) {
+                    const index = response.asyncFunctionArgs[0]
+                    nextState = { ...response.state, stack: [...(response.state.stack || []), null] }
+                    actions.setExecResult(index, {
+                        stdout: stdout.join('\n'),
+                        state: nextState,
+                    })
+                    stdout = []
+                } else {
+                    lemonToast.error(`Unknown async function: ${response.asyncFunctionName}`)
+                    break
+                }
+            }
+        },
+    })),
+    actionToUrl(({ values }) => ({
+        updateCode: () => ['/hogbooks', {}, { codes: values.codes }],
+    })),
+    urlToAction(({ actions, values }) => ({
+        '/hogbooks': (_, __, { codes }) => {
+            if (codes && !equal(codes, values.codes)) {
+                actions.updateCodes(codes)
+            }
+        },
+    })),
+])
+
+export function NotebookTest(): JSX.Element {
+    const { codes, outputs, bytecodeUntil, runUntil } = useValues(notebookTestLogic)
+    const { runFrom, updateCode } = useActions(notebookTestLogic)
+    const blocks = 3
+    return (
+        <div className="NotebookTest space-y-2">
```

Comment:
I get this is just a demo but...

The real "power" here would be making it that any Query could output some sort of generic "input,outputs,refresh" interface. That way you could take the data for a hogql query, the data from an insight, even a list of recordings and use them as an input to the next NotebookNode. Hog at that point is just a simple node where you can transform that data in anyway you would desire.

Reply:
For sure. I've been thinking of it and even have a work in progress that's too raw to share. Basically I need to make the compiler itself "resumable"... and I think it should be as easy as just passing on `locals` in addition to `bytecode` from the compiler. This `locals` is an array of strings with names of the local variables in the root scope. Coincidentally the array of locals will map 1:1 onto the `stack` array.